### ISSUE TITLE
Packaging 4.08.0 on Arch Linux fails

### ISSUE THREAD

**Issue by juergenhoetzel** [ocaml/ocaml#8742]
Trying to build the `4.08.0` release using this [PKGBUILD](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/ocaml)
results in:
```
==> Making package: ocaml 4.08.0-1 (Sun Jun 16 18:21:40 2019)
...
==> Starting package_ocaml()...
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```
the build step passes but packaging fails. I used `bcc` to trace the `git` command:
```
[root@lemmy ~]# execsnoop.py -n git
PCOMM            PID    PPID   RET ARGS
git              24918  24917    0 /usr/bin/git ls-files *.ml *.mli

```
I wonder why git is used in the release tarball at all.

Arch Linux uses [fakeroot](http://packages.debian.org/fakeroot) to build a package.  I suppose this might be one reason for the build error.

any help is appreciated :bowing_man:

**Comment by dra27** [comment]
This is an issue with #2030. I'll formulate a proper fix shortly, but for 4.08.0 you're going to need to patch the sources. I think this should fix it:

```diff
diff --git a/Makefile b/Makefile
index 2a9413a97a..ec9be30a30 100644
--- a/Makefile
+++ b/Makefile
@@ -1305,9 +1305,6 @@ CAMLC_DPARSETREE := \
 list-all-asts:
        @for f in $(AST_FILES); do echo "'$$f'"; done
 
-partialclean::
-       rm -f $(AST_FILES)
-
 # Default rules
 
 .SUFFIXES: .ml .mli .cmo .cmi .cmx
```

Please could you confirm?

**Comment by dra27** [comment]
(I've tweaked the title only to try to make it clear that the problem is only when building with the package manager on Arch - I would think that building OCaml from a git clone, or manually from a tarball would succeed)

**Comment by juergenhoetzel** [comment]
Thanks for the quick response :+1: 
> Please could you confirm?

sadly not.
I don't understand why removing the `partialclean` target should fix the error.

**Comment by dra27** [comment]
When you say "sadly not", you mean you can't confirm it, or you can confirm that it _doesn't_ fix it? With that patch it's definitely still running `git ls-files *.ml *.mli`?

**Comment by dra27** [comment]
(incidentally, looking at the `PKGBUILD`, I don't think you need any of `gdbm`, `tk` or `ncurses`)

**Comment by juergenhoetzel** [comment]
> When you say "sadly not", you mean you can't confirm it, or you can confirm that it _doesn't_ fix it? With that patch it's definitely still running `git ls-files *.ml *.mli`?

The patch doesn't fix it for me. It's still running `git` multiple times (also during the build process):
```
git              20918  20917    0 /usr/bin/git ls-files *.ml *.mli
git              21177  21176    0 /usr/bin/git ls-files *.ml *.mli
git              22408  22407    0 /usr/bin/git ls-files *.ml *.mli
git              22703  22702    0 /usr/bin/git ls-files *.ml *.mli
git              22810  22809    0 /usr/bin/git ls-files *.ml *.mli
git              23941  23940    0 /usr/bin/git ls-files *.ml *.mli
```

> I would think that building OCaml from a git clone, or manually from a tarball would succeed)

Yes: Building/Installing without the Arch Linux build tools (`makepkg`, `extra-x86_64-build`) succeeds.

**Comment by dra27** [comment]
Is it possible to get a full log of the build? My original hunch was that it was the `make partialclean` (since that should be the only place where `$(SOURCE_FILES)`, which is where the `git ls-files` call comes from, is forced), but I forgot that that only occurs during a bootstrap or flexdll bootstrap, which shouldn't apply here. I'm wondering if it's dumping make variables at some point, but I think the only way to track it is to see the complete log.

**Comment by juergenhoetzel** [comment]
> Is it possible to get a full log of the build?

Of course:
[build.log](https://github.com/ocaml/ocaml/files/3298231/build.log)
Using `make --debug=v`:
[build-verbose.log](https://github.com/ocaml/ocaml/files/3298232/build-verbose.log)

**Comment by dra27** [comment]
Thanks - are those the build logs with the patch I suggested or from before?

<pre><font color="#8AE234"><b>==&gt;</b></font><b> Entering fakeroot environment...</b>
<font color="#8AE234"><b>==&gt;</b></font><b> Starting package_ocaml()...</b>
/bin/sh: git: command not found
mkdir -p &quot;/usr/bin&quot;
mkdir -p &quot;/usr/lib/ocaml&quot;
mkdir: cannot create directory ‘/usr/lib/ocaml’: Permission denied
make: *** [Makefile:517: install] Error 1
</pre>

**Comment by juergenhoetzel** [comment]
> Thanks - are those the build logs with the patch I suggested or from before?

The build didn't contain the patch. But I get the same `git: command not found` error when the patch is applied. Anyway: The error is non-fatal, I should have read the error output more closely:

```
> ==> Entering fakeroot environment...
> ==> Starting package_ocaml()...
> /bin/sh: git: command not found
> mkdir -p "/usr/bin"
> mkdir -p "/usr/lib/ocaml"
> mkdir: cannot create directory /usr/lib/ocaml: Permission denied
```

The `install` target doesn't install to the  buildroot:

```
make PREFIX="${pkgdir}/usr" install
```

The previous Arch Linux maintainer used `PREFIX` instead of `DESTDIR` (I don't know why). With version 4.07.1 this still worked. 

Using [DESTDIR](https://git.archlinux.org/svntogit/packages.git/commit/trunk?h=packages/ocaml&id=5850aad0a69a833b151246524492e4e9756805c7) the install target succeeds (apart from the git error message).

**Comment by juergenhoetzel** [comment]
>```
> [root@lemmy ~]# execsnoop.py -n git
> PCOMM            PID    PPID   RET ARGS
> git              24918  24917    0 /usr/bin/git ls-files *.ml *.mli
> ```
Couldn't this be replaced by
```
find . -name "*.ml" -o -name "*mli"
```

so the tarball release doesn't depend on `git`?

The reason why the build continues (despite this missing `git` command) seems the absence of `set -o pipefail`:

```
SOURCE_FILES=$(shell git ls-files '*.ml' '*.mli' | grep -v boot/menhir/parser)
```

**Comment by dra27** [comment]
What's puzzling me is why that variable is being evaluated at all - it shouldn't be. `set -o pipefail` is a bashism, I think?

There are various solutions, but I'd like first to understand why a variable which should not be evaluated is being evaluated at all (it's defined with `=`, not `:=`, and Posix is fairly clear on the subject!)

**Comment by Armael** [comment]
I'm a complete make noob, but doesn't [this line](https://github.com/ocaml/ocaml/blob/trunk/Makefile#L1276) (also present in the 4.08 makefile) force make to evaluate `$(AST_FILES)` in all cases? (and by transitivity, evaluating `$(SOURCE_FILES)` which runs git)

**Comment by dra27** [comment]
@Armael - yes, that’s right, thanks! It evaluates dependencies even if you don’t request the rule. I’ll amend the PR accordingly.

**Comment by juergenhoetzel** [comment]
> I'm a complete make noob, but doesn't [this line](https://github.com/ocaml/ocaml/blob/trunk/Makefile#L1276) (also present in the 4.08 makefile) force make to evaluate `$(AST_FILES)` in all cases? (and by transitivity, evaluating `$(SOURCE_FILES)` which runs git)

I can confirm: Removing this rule solves the problem. Thanks! :+1:

**Comment by dra27** [comment]
I've pushed an additional fix to #8743, so this should be fixed in 4.08.1+4.09.0

**Comment by dra27** [comment]
This has been fixed by moving the rules in question to `Makefile.dev` which is not `include`d if there is no `.git` file/directory in the build tree.
